modules/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,467 @@
+    }
+
+    /* Reset line */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWire, ROM);
+    /* Read scratchpad command by onewire protocol */
+    OneWire_WriteByte(OneWire, ONEWIRE_CMD_RSCRATCHPAD);
+    
+    /* Get data */
+    for (i = 0; i < DS18B20_READ_DATA_SIZE; i++) {
+        /* Read byte by byte */
+        OneWire_ReadByte(OneWire, &data[i]);
+    }
+    
+    /* Calculate CRC */
+    // OneWire_CRC8(data, 8, &crc);
+    crc = OneWire_LookupCRC8(data,8);
+
+    /* Check if CRC is ok */
+    if (crc != data[DS18B20_READ_CRC_BYTE]) {
+        /* CRC invalid */
+        return DS18B20_FAILURE;
+    }
+    
+    /* First two bytes of scratchpad are temperature values */
+    temperature = data[DS18B20_DATA_LSB] | (data[DS18B20_DATA_MSB] << DS18B20_READ_CRC_BYTE);
+
+    /* Reset line */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    
+    /* Check if temperature is negative */
+    if (temperature & DS18B20_TEMP_SIGN_BITMASK) {
+        /* Two's complement, temperature is negative */
+        temperature = ~temperature + 1;
+        minus = true;
+    }
+
+    /* Get sensor resolution */
+    resolution = DS18B20_ConfigRegToResolution(data[DS18B20_CONFIG_REGISTER_BYTE]);
+    
+
+    /* Measure temperature by first extracting the digit (non decimal) portion. */
+    /* Ignore all but 1 signed bit */
+    digit = temperature >> 4;
+    /* extract signed bit and 3 digit temp bits from MSB, put in appropriate power */
+    digit |= ((temperature >> 8) & 0x7) << 4;
+    
+    /* Store decimal digits */
+    /* Resolution may be either 9, 10, 11, or 12 bits.
+    If the DS18B20  is  configured for  12-bit resolution, all bits in the temperature
+    register will contain valid  data. For 11-bit resolution, bit 0 is undefined. 
+    For  10-bit resolution, bits 1 and 0 are undefined, and for 9-bit resolution  
+    bits  2,  1,  and  0  are  undefined. Shift as needed to extract 1 step of 
+    the decimal temperature then multiply by decimal steps per resolution */
+    switch (resolution) {
+        case DS18B20_RESOLUTION_9BITS: {
+            decimal = (temperature >> 3) & 0x01;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_9BIT;
+            break;
+        } 
+        case DS18B20_RESOLUTION_10BITS: {
+            decimal = (temperature >> 2) & 0x03;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_10BIT;
+            break;
+        }
+        case DS18B20_RESOLUTION_11BITS: {
+            decimal = (temperature >> 1) & 0x07;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_11BIT;
+            break;
+        } 
+        case DS18B20_RESOLUTION_12BITS: {
+            decimal = temperature & 0x0F;
+            decimal *= (float)DS18B20_DECIMAL_STEPS_12BIT;
+            break;
+        }
+        default: {
+            decimal = 0xFF;
+            digit = 0;
+            break;

```

Comment:
I'm assuming the expectation is that we'd never hit the default case and that if we did then something went horribly wrong.  So if we do hit the default case, should we return some kind of DS18B20 error status?

Reply:
This is technically impossible since setting a non DS18B20 9, 10, 11, or 12 value is forbidden, but I agree this should be taken care of.